### PR TITLE
Improvements to types and testing

### PR DESCRIPTION
**Pull request by sp3ctum**
Here are some updates that don't add new features, but should make maintaining the project and contributing to it that much easier:

# refactor(types): add explicit types
No runtime change, but maybe slightly easier to understand and maintain the code.

---

# chore: all development dependencies are locked and reliably installable
This is a change that builds on the base of the build system I worked on a few days ago. It adds the following improvements:
- the exact versions of all testing dependencies are stored within the project (previously they were outside).
  - the versions are stored in a  lock file. I used https://github.com/folke/lazy.nvim because it's the only modern one I know.
  - lazy.nvim is used both to install and maintain the dependencies, as well as loading them when it's time to test.

---

# feat(development): controlled playground nvim instance
This is a side effect of the previous one. The testing setup can also be used to launch a minimal nvim that can be used to try the plugin out.
- idea: in the future, it could be used to reproduce bugs or collaborate with developers

Here is a mini demo. In the demo, I first install dependencies (according to the lock file) and then run tests.

https://github.com/gsuuon/tshjkl.nvim/assets/300791/b534cc78-d3f6-48a5-91e8-8045fb7d7b69

---

# docs: contribution instructions
These are pretty basic but should cover the essentials.

### CODE REVIEW

**Comment by sp3ctum** [comment]
I would really like to hear what you think about these ideas, especially if you'd like to take the project in another direction in some way.

I wasn't able to find a similar/simple example in the nvim plugin developer community, so I hacked something together myself 😄 

I think in general, a simple "how to get started" experience would be good, but I am not sure the specifics are good fits, etc.

**Comment by sp3ctum** [comment]
Hmm looks like I forgot to make the test environment initialization quit neovim - now it probably can't proceed automatically. I can fix this later.

**Comment by gsuuon** [comment]
Hey sorry I've been a bit busy lately, I haven't forgotten about these and I'll take a look this weekend.

**Comment by gsuuon** [comment]
This project is meant to be a really small, simple plugin - I think this overcomplicates setup. Plenary is the only testing dependency. We don't need nvim-treesitter or tree-sitter-lua as dependencies since neovim ships with treesitter and a lua parser. This plugin doesn't ever do anything language specific (we won't do queries) so the built-in stuff is all we'll ever need for testing.

I don't think a contributing guide is necessary in this project since there shouldn't be any dev setup. I do think the types will be useful here, but the rest are overkill for this project. 

For plugin examples, I usually take a look at the bigger neovim plugins (like lazy or telescope) and check their github workflows. I think the ideas in this PR would be good to explore in a separate experimental repo for neovim dev things, though for consistency I'd lean more towards a dev container than lazy + lockfile (so you can pin the neovim version as well). It's not generally necessary though since neovim plugins don't have deep dependency trees.

File: lua/tshjkl/init.lua [hunk 1]
```diff
@@ -354,8 +397,11 @@ local function enter(outermost)
 end
 
 local function keybind_global(opts)
+  ---@param outermost boolean
+  ---@return fun(): nil
   local function toggle(outermost)
     return function()
+      print('HELLO WORLD')
```
Comment (gsuuon) [inline lua/tshjkl/init.lua:404]: Make sure to strip this out as well!

**Review by gsuuon — CHANGES_REQUESTED** [review]
Lets just stick to adding types here

**Comment by sp3ctum** [comment]
> Lets just stick to adding types here

I made some changes and only left the type additions in.

**Review by gsuuon — APPROVED** [review]
Thanks!

**Pull request merged**